// unleash-server/dist/lib/middleware/rbac-middleware.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const rbac_middleware_1 = __importDefault(require("./rbac-middleware"));
const fake_feature_toggle_store_1 = __importDefault(require("../../test/fixtures/fake-feature-toggle-store"));
const user_1 = __importDefault(require("../types/user"));
const perms = __importStar(require("../types/permissions"));
const test_config_1 = require("../../test/config/test-config");
const api_user_1 = __importDefault(require("../types/api-user"));
let config;
let featureToggleStore;
beforeEach(() => {
    featureToggleStore = fake_feature_toggle_store_1.default();
    config = test_config_1.createTestConfig();
});
test('should add checkRbac to request', () => {
    const accessService = {
        hasPermission: jest.fn(),
    };
    const func = rbac_middleware_1.default(config, { featureToggleStore }, accessService);
    const cb = jest.fn();
    const req = jest.fn();
    func(req, undefined, cb);
    // @ts-ignore
    expect(req.checkRbac).toBeTruthy();
    // @ts-ignore
    expect(typeof req.checkRbac).toBe('function');
});
test('should give api-user ADMIN permission', async () => {
    const accessService = {
        hasPermission: jest.fn(),
    };
    const func = rbac_middleware_1.default(config, { featureToggleStore }, accessService);
    const cb = jest.fn();
    const req = {
        user: new api_user_1.default({
            username: 'api',
            permissions: [perms.ADMIN],
        }),
    };
    func(req, undefined, cb);
    const hasAccess = await req.checkRbac(perms.ADMIN);
    expect(hasAccess).toBe(true);
});
test('should not give api-user ADMIN permission', async () => {
    const accessService = {
        hasPermission: jest.fn(),
    };
    const func = rbac_middleware_1.default(config, { featureToggleStore }, accessService);
    const cb = jest.fn();
    const req = {
        user: new api_user_1.default({
            username: 'api',
            permissions: [perms.CLIENT],
        }),
    };
    func(req, undefined, cb);
    const hasAccess = await req.checkRbac(perms.ADMIN);
    expect(hasAccess).toBe(false);
    expect(accessService.hasPermission).toHaveBeenCalledTimes(0);
});
test('should not allow user to miss userId', async () => {
    const accessService = {
        hasPermission: jest.fn(),
    };
    const func = rbac_middleware_1.default(config, { featureToggleStore }, accessService);
    const cb = jest.fn();
    const req = {
        user: {
            username: 'user',
        },
    };
    func(req, undefined, cb);
    const hasAccess = await req.checkRbac(perms.ADMIN);
    expect(hasAccess).toBe(false);
});
test('should return false for missing user', async () => {
    const accessService = {
        hasPermission: jest.fn(),
    };
    const func = rbac_middleware_1.default(config, { featureToggleStore }, accessService);
    const cb = jest.fn();
    const req = {};
    func(req, undefined, cb);
    const hasAccess = await req.checkRbac(perms.ADMIN);
    expect(hasAccess).toBe(false);
    expect(accessService.hasPermission).toHaveBeenCalledTimes(0);
});
test('should verify permission for root resource', async () => {
    const accessService = {
        hasPermission: jest.fn(),
    };
    const func = rbac_middleware_1.default(config, { featureToggleStore }, accessService);
    const cb = jest.fn();
    const req = {
        user: new user_1.default({
            username: 'user',
            id: 1,
        }),
        params: {},
    };
    func(req, undefined, cb);
    await req.checkRbac(perms.ADMIN);
    expect(accessService.hasPermission).toHaveBeenCalledTimes(1);
    expect(accessService.hasPermission).toHaveBeenCalledWith(req.user, perms.ADMIN, undefined);
});
test('should lookup projectId from params', async () => {
    const accessService = {
        hasPermission: jest.fn(),
    };
    const func = rbac_middleware_1.default(config, { featureToggleStore }, accessService);
    const cb = jest.fn();
    const req = {
        user: new user_1.default({
            username: 'user',
            id: 1,
        }),
        params: {
            projectId: 'some-proj',
        },
    };
    func(req, undefined, cb);
    await req.checkRbac(perms.UPDATE_PROJECT);
    expect(accessService.hasPermission).toHaveBeenCalledWith(req.user, perms.UPDATE_PROJECT, req.params.projectId);
});
test('should lookup projectId from feature toggle', async () => {
    const projectId = 'some-project-33';
    const featureName = 'some-feature-toggle';
    const accessService = {
        hasPermission: jest.fn(),
    };
    featureToggleStore.getProjectId = jest.fn().mockReturnValue(projectId);
    const func = rbac_middleware_1.default(config, { featureToggleStore }, accessService);
    const cb = jest.fn();
    const req = {
        user: new user_1.default({
            username: 'user',
            id: 1,
        }),
        params: {
            featureName,
        },
    };
    func(req, undefined, cb);
    await req.checkRbac(perms.UPDATE_FEATURE);
    expect(accessService.hasPermission).toHaveBeenCalledWith(req.user, perms.UPDATE_FEATURE, projectId);
    expect(featureToggleStore.getProjectId.mock.calls[0][0]).toBe(featureName);
});
test('should lookup projectId from data', async () => {
    const projectId = 'some-project-33';
    const featureName = 'some-feature-toggle';
    const accessService = {
        hasPermission: jest.fn(),
    };
    const func = rbac_middleware_1.default(config, { featureToggleStore }, accessService);
    const cb = jest.fn();
    const req = {
        user: new user_1.default({
            username: 'user',
            id: 1,
        }),
        params: {},
        body: {
            featureName,
            project: projectId,
        },
    };
    func(req, undefined, cb);
    await req.checkRbac(perms.CREATE_FEATURE);
    expect(accessService.hasPermission).toHaveBeenCalledWith(req.user, perms.CREATE_FEATURE, projectId);
});
test('Need access to UPDATE_FEATURE on the project you change to', async () => {
    const oldProjectId = 'some-project-34';
    const newProjectId = 'some-project-35';
    const featureName = 'some-feature-toggle';
    const accessService = {
        hasPermission: jest.fn().mockReturnValue(true),
    };
    featureToggleStore.getProjectId = jest.fn().mockReturnValue(oldProjectId);
    const func = rbac_middleware_1.default(config, { featureToggleStore }, accessService);
    const cb = jest.fn();
    const req = {
        user: new user_1.default({ username: 'user', id: 1 }),
        params: { featureName },
        body: { featureName, project: newProjectId },
    };
    func(req, undefined, cb);
    await req.checkRbac(perms.UPDATE_FEATURE);
    expect(accessService.hasPermission).toHaveBeenCalledTimes(2);
    expect(accessService.hasPermission).toHaveBeenNthCalledWith(1, req.user, perms.UPDATE_FEATURE, oldProjectId);
    expect(accessService.hasPermission).toHaveBeenNthCalledWith(2, req.user, perms.UPDATE_FEATURE, newProjectId);
});
test('Does not double check permission if not changing project when updating toggle', async () => {
    const oldProjectId = 'some-project-34';
    const featureName = 'some-feature-toggle';
    const accessService = {
        hasPermission: jest.fn().mockReturnValue(true),
    };
    featureToggleStore.getProjectId = jest.fn().mockReturnValue(oldProjectId);
    const func = rbac_middleware_1.default(config, { featureToggleStore }, accessService);
    const cb = jest.fn();
    const req = {
        user: new user_1.default({ username: 'user', id: 1 }),
        params: { featureName },
        body: { featureName, project: oldProjectId },
    };
    func(req, undefined, cb);
    await req.checkRbac(perms.UPDATE_FEATURE);
    expect(accessService.hasPermission).toHaveBeenCalledTimes(1);
    expect(accessService.hasPermission).toHaveBeenCalledWith(req.user, perms.UPDATE_FEATURE, oldProjectId);
});
