// unleash-server/dist/lib/db/context-field-store.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const COLUMNS = [
    'name',
    'description',
    'stickiness',
    'sort_order',
    'legal_values',
    'created_at',
];
const TABLE = 'context_fields';
const mapRow = row => ({
    name: row.name,
    description: row.description,
    stickiness: row.stickiness,
    sortOrder: row.sort_order,
    legalValues: row.legal_values ? row.legal_values.split(',') : undefined,
    createdAt: row.created_at,
});
class ContextFieldStore {
    constructor(db, getLogger) {
        this.db = db;
        this.logger = getLogger('context-field-store.js');
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    fieldToRow(data) {
        return {
            name: data.name,
            description: data.description,
            stickiness: data.stickiness,
            sort_order: data.sortOrder,
            legal_values: data.legalValues ? data.legalValues.join(',') : null,
            updated_at: data.createdAt, // eslint-disable-line
        };
    }
    async getAll() {
        const rows = await this.db
            .select(COLUMNS)
            .from(TABLE)
            .orderBy('name', 'asc');
        return rows.map(mapRow);
    }
    async get(name) {
        return this.db
            .first(COLUMNS)
            .from(TABLE)
            .where({ name })
            .then(mapRow);
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    async create(contextField) {
        return this.db(TABLE).insert(this.fieldToRow(contextField));
    }
    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
    async update(data) {
        return this.db(TABLE)
            .where({ name: data.name })
            .update(this.fieldToRow(data));
    }
    async delete(name) {
        return this.db(TABLE)
            .where({ name })
            .del();
    }
}
exports.default = ContextFieldStore;
module.exports = ContextFieldStore;
