// unleash-server/dist/lib/db/session-store.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const notfound_error_1 = __importDefault(require("../error/notfound-error"));
const TABLE = 'unleash_session';
class SessionStore {
    constructor(db, eventBus, getLogger) {
        this.db = db;
        this.eventBus = eventBus;
        this.logger = getLogger('lib/db/session-store.ts');
    }
    async getActiveSessions() {
        const rows = await this.db(TABLE)
            .whereNull('expired')
            .orWhere('expired', '>', new Date())
            .orderBy('created_at', 'desc');
        return rows.map(this.rowToSession);
    }
    async getSessionsForUser(userId) {
        const rows = await this.db(TABLE).whereRaw(`(sess -> 'user' ->> 'id')::int = ?`, [userId]);
        if (rows && rows.length > 0) {
            return rows.map(this.rowToSession);
        }
        throw new notfound_error_1.default(`Could not find sessions for user with id ${userId}`);
    }
    async getSession(sid) {
        const row = await this.db(TABLE)
            .where('sid', '=', sid)
            .first();
        if (row) {
            return this.rowToSession(row);
        }
        throw new notfound_error_1.default(`Could not find session with sid ${sid}`);
    }
    async deleteSessionsForUser(userId) {
        await this.db(TABLE)
            .whereRaw(`(sess -> 'user' ->> 'id')::int = ?`, [userId])
            .del();
    }
    async deleteSession(sid) {
        await this.db(TABLE)
            .where('sid', '=', sid)
            .del();
    }
    async insertSession(data) {
        const row = await this.db(TABLE)
            .insert({
            sid: data.sid,
            sess: JSON.stringify(data.sess),
            expired: data.expired || new Date(Date.now() + 86400000),
        })
            .returning(['sid', 'sess', 'created_at', 'expired']);
        if (row) {
            return this.rowToSession(row);
        }
        throw new Error('Could not insert session');
    }
    async deleteAll() {
        await this.db(TABLE).del();
    }
    rowToSession(row) {
        return {
            sid: row.sid,
            sess: row.sess,
            createdAt: row.created_at,
            expired: row.expired,
        };
    }
}
exports.default = SessionStore;
module.exports = SessionStore;
