// unleash-server/dist/lib/db/user-store.js
"use strict";
/* eslint camelcase: "off" */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const user_1 = __importDefault(require("../types/user"));
const NotFoundError = require('../error/notfound-error');
const TABLE = 'users';
const USER_COLUMNS = [
    'id',
    'name',
    'username',
    'email',
    'image_url',
    'login_attempts',
    'seen_at',
    'created_at',
];
const USER_COLUMNS_PUBLIC = ['id', 'name', 'username', 'email', 'image_url'];
const emptify = value => {
    if (!value) {
        return undefined;
    }
    return value;
};
const safeToLower = (s) => (s ? s.toLowerCase() : s);
const mapUserToColumns = (user) => ({
    name: user.name,
    username: user.username,
    email: safeToLower(user.email),
    image_url: user.imageUrl,
});
const rowToUser = row => {
    if (!row) {
        throw new NotFoundError('No user found');
    }
    return new user_1.default({
        id: row.id,
        name: emptify(row.name),
        username: emptify(row.username),
        email: emptify(row.email),
        imageUrl: emptify(row.image_url),
        loginAttempts: row.login_attempts,
        seenAt: row.seen_at,
        createdAt: row.created_at,
    });
};
class UserStore {
    constructor(db, getLogger) {
        this.db = db;
        this.logger = getLogger('user-store.js');
    }
    async update(id, fields) {
        await this.db(TABLE)
            .where('id', id)
            .update(mapUserToColumns(fields));
        return this.get({ id });
    }
    async insert(user) {
        const rows = await this.db(TABLE)
            .insert(mapUserToColumns(user))
            .returning(USER_COLUMNS);
        return rowToUser(rows[0]);
    }
    async upsert(user) {
        const id = await this.hasUser(user);
        if (id) {
            return this.update(id, user);
        }
        return this.insert(user);
    }
    buildSelectUser(q) {
        const query = this.db(TABLE);
        if (q.id) {
            return query.where('id', q.id);
        }
        if (q.email) {
            return query.where('email', safeToLower(q.email));
        }
        if (q.username) {
            return query.where('username', q.username);
        }
        throw new Error('Can only find users with id, username or email.');
    }
    async hasUser(idQuery) {
        const query = this.buildSelectUser(idQuery);
        const item = await query.first('id');
        return item ? item.id : undefined;
    }
    async getAll() {
        const users = await this.db.select(USER_COLUMNS).from(TABLE);
        return users.map(rowToUser);
    }
    async search(query) {
        const users = await this.db
            .select(USER_COLUMNS_PUBLIC)
            .from(TABLE)
            .where('name', 'ILIKE', `%${query}%`)
            .orWhere('username', 'ILIKE', `${query}%`)
            .orWhere('email', 'ILIKE', `${query}%`);
        return users.map(rowToUser);
    }
    async getAllWithId(userIdList) {
        const users = await this.db
            .select(USER_COLUMNS_PUBLIC)
            .from(TABLE)
            .whereIn('id', userIdList);
        return users.map(rowToUser);
    }
    async get(idQuery) {
        const row = await this.buildSelectUser(idQuery).first(USER_COLUMNS);
        return rowToUser(row);
    }
    async delete(id) {
        return this.db(TABLE)
            .where({ id })
            .del();
    }
    async getPasswordHash(userId) {
        const item = await this.db(TABLE)
            .where('id', userId)
            .first('password_hash');
        if (!item) {
            throw new NotFoundError('User not found');
        }
        return item.password_hash;
    }
    async setPasswordHash(userId, passwordHash) {
        return this.db(TABLE)
            .where('id', userId)
            .update({
            password_hash: passwordHash,
        });
    }
    async incLoginAttempts(user) {
        return this.buildSelectUser(user).increment('login_attempts', 1);
    }
    async successfullyLogin(user) {
        return this.buildSelectUser(user).update({
            login_attempts: 0,
            seen_at: new Date(),
        });
    }
    async deleteAll() {
        await this.db(TABLE).del();
    }
}
module.exports = UserStore;
exports.default = UserStore;
