// unleash-server/dist/server-dev.js
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const server_impl_1 = __importDefault(require("./lib/server-impl"));
const create_config_1 = require("./lib/create-config");
const logger_1 = require("./lib/logger");
process.nextTick(async () => {
    try {
        await server_impl_1.default.start(create_config_1.createConfig({
            db: {
                user: 'unleash_user',
                password: 'passord',
                host: 'localhost',
                port: 5432,
                database: 'unleash',
                ssl: false,
            },
            server: {
                enableRequestLogger: true,
                baseUriPath: '',
                // keepAliveTimeout: 1,
                gracefulShutdownEnable: true,
            },
            logLevel: logger_1.LogLevel.debug,
            enableOAS: true,
            versionCheck: {
                enable: false,
            },
        }));
    }
    catch (error) {
        if (error.code === 'EADDRINUSE') {
            // eslint-disable-next-line no-console
            console.warn('Port in use. You might want to reload once more.');
        }
        else {
            // eslint-disable-next-line no-console
            console.error(error);
            process.exit();
        }
    }
}, 0);
