// unleash-server/dist/lib/services/access-service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccessService = exports.RoleType = exports.RoleName = exports.ALL_PROJECTS = void 0;
const permissions = __importStar(require("../types/permissions"));
exports.ALL_PROJECTS = '*';
const PROJECT_DESCRIPTION = {
    OWNER: 'Users with this role have full control over the project, and can add and manage other users within the project context, manage feature toggles within the project, and control advanced project features like archiving and deleting the project.',
    MEMBER: 'Users with this role within a project are allowed to view, create and update feature toggles, but have limited permissions in regards to managing the projects user access and can not archive or delete the project.',
};
const { ADMIN } = permissions;
const PROJECT_ADMIN = [
    permissions.UPDATE_PROJECT,
    permissions.DELETE_PROJECT,
    permissions.CREATE_FEATURE,
    permissions.UPDATE_FEATURE,
    permissions.DELETE_FEATURE,
];
const PROJECT_REGULAR = [
    permissions.CREATE_FEATURE,
    permissions.UPDATE_FEATURE,
    permissions.DELETE_FEATURE,
];
const isProjectPermission = permission => PROJECT_ADMIN.includes(permission);
var PermissionType;
(function (PermissionType) {
    PermissionType["root"] = "root";
    PermissionType["project"] = "project";
})(PermissionType || (PermissionType = {}));
var RoleName;
(function (RoleName) {
    // eslint-disable-next-line @typescript-eslint/no-shadow
    RoleName["ADMIN"] = "Admin";
    RoleName["EDITOR"] = "Editor";
    RoleName["VIEWER"] = "Viewer";
    RoleName["OWNER"] = "Owner";
    RoleName["MEMBER"] = "Member";
})(RoleName = exports.RoleName || (exports.RoleName = {}));
var RoleType;
(function (RoleType) {
    RoleType["ROOT"] = "root";
    RoleType["PROJECT"] = "project";
})(RoleType = exports.RoleType || (exports.RoleType = {}));
class AccessService {
    constructor({ accessStore, userStore }, { getLogger }) {
        this.RoleName = RoleName;
        this.store = accessStore;
        this.userStore = userStore;
        this.logger = getLogger('/services/access-service.ts');
        this.permissions = Object.values(permissions).map(p => ({
            name: p,
            type: isProjectPermission(p)
                ? PermissionType.project
                : PermissionType.root,
        }));
    }
    /**
     * Used to check if a user has access to the requested resource
     *
     * @param user
     * @param permission
     * @param projectId
     */
    async hasPermission(user, permission, projectId) {
        this.logger.info(`Checking permission=${permission}, userId=${user.id} projectId=${projectId}`);
        try {
            const userP = await this.store.getPermissionsForUser(user.id);
            return userP
                .filter(p => !p.project ||
                p.project === projectId ||
                p.project === exports.ALL_PROJECTS)
                .some(p => p.permission === permission || p.permission === ADMIN);
        }
        catch (e) {
            this.logger.error(`Error checking permission=${permission}, userId=${user.id} projectId=${projectId}`, e);
            return Promise.resolve(false);
        }
    }
    async getPermissionsForUser(user) {
        if (user.isAPI) {
            return [];
        }
        return this.store.getPermissionsForUser(user.id);
    }
    getPermissions() {
        return this.permissions;
    }
    async addUserToRole(userId, roleId) {
        return this.store.addUserToRole(userId, roleId);
    }
    async setUserRootRole(userId, role) {
        const newRootRole = await this.resolveRootRole(role);
        if (newRootRole) {
            try {
                await this.store.removeRolesOfTypeForUser(userId, RoleType.ROOT);
                await this.store.addUserToRole(userId, newRootRole.id);
            }
            catch (error) {
                throw new Error(`Could not add role=${newRootRole.name} to userId=${userId}`);
            }
        }
        else {
            throw new Error(`Could not find rootRole=${role}`);
        }
    }
    async getUserRootRoles(userId) {
        const userRoles = await this.store.getRolesForUserId(userId);
        return userRoles.filter(r => r.type === RoleType.ROOT);
    }
    async removeUserFromRole(userId, roleId) {
        return this.store.removeUserFromRole(userId, roleId);
    }
    async addPermissionToRole(roleId, permission, projectId) {
        if (isProjectPermission(permission) && !projectId) {
            throw new Error(`ProjectId cannot be empty for permission=${permission}`);
        }
        return this.store.addPermissionsToRole(roleId, [permission], projectId);
    }
    async removePermissionFromRole(roleId, permission, projectId) {
        if (isProjectPermission(permission) && !projectId) {
            throw new Error(`ProjectId cannot be empty for permission=${permission}`);
        }
        return this.store.removePermissionFromRole(roleId, permission, projectId);
    }
    async getRoles() {
        return this.store.getRoles();
    }
    async getRole(roleId) {
        const [role, rolePerms, users] = await Promise.all([
            this.store.getRoleWithId(roleId),
            this.store.getPermissionsForRole(roleId),
            this.getUsersForRole(roleId),
        ]);
        return { role, permissions: rolePerms, users };
    }
    async getRolesForProject(projectId) {
        return this.store.getRolesForProject(projectId);
    }
    async getRolesForUser(userId) {
        return this.store.getRolesForUserId(userId);
    }
    async getUsersForRole(roleId) {
        const userIdList = await this.store.getUserIdsForRole(roleId);
        return this.userStore.getAllWithId(userIdList);
    }
    // Move to project-service?
    async getProjectRoleUsers(projectId) {
        const roles = await this.store.getRolesForProject(projectId);
        const users = await Promise.all(roles.map(async (role) => {
            const usrs = await this.getUsersForRole(role.id);
            return usrs.map(u => ({ ...u, roleId: role.id }));
        }));
        return [roles, users.flat()];
    }
    async createDefaultProjectRoles(owner, projectId) {
        if (!projectId) {
            throw new Error('ProjectId cannot be empty');
        }
        const ownerRole = await this.store.createRole(RoleName.OWNER, RoleType.PROJECT, projectId, PROJECT_DESCRIPTION.OWNER);
        await this.store.addPermissionsToRole(ownerRole.id, PROJECT_ADMIN, projectId);
        // TODO: remove this when all users is guaranteed to have a unique id.
        if (owner.id) {
            this.logger.info(`Making ${owner.id} admin of ${projectId} via roleId=${ownerRole.id}`);
            await this.store.addUserToRole(owner.id, ownerRole.id);
        }
        const memberRole = await this.store.createRole(RoleName.MEMBER, RoleType.PROJECT, projectId, PROJECT_DESCRIPTION.MEMBER);
        await this.store.addPermissionsToRole(memberRole.id, PROJECT_REGULAR, projectId);
    }
    async removeDefaultProjectRoles(owner, projectId) {
        this.logger.info(`Removing project roles for ${projectId}`);
        return this.store.removeRolesForProject(projectId);
    }
    async getRootRoleForAllUsers() {
        return this.store.getRootRoleForAllUsers();
    }
    async getRootRoles() {
        return this.store.getRootRoles();
    }
    async resolveRootRole(rootRole) {
        const rootRoles = await this.getRootRoles();
        let role;
        if (typeof rootRole === 'number') {
            role = rootRoles.find(r => r.id === rootRole);
        }
        else {
            role = rootRoles.find(r => r.name === rootRole);
        }
        return role;
    }
    async getRootRole(roleName) {
        const roles = await this.store.getRootRoles();
        return roles.find(r => r.name === roleName);
    }
}
exports.AccessService = AccessService;
