// unleash-server/dist/lib/routes/admin-api/config.js
'use strict';
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const version_1 = __importDefault(require("../../util/version"));
const Controller = require('../controller');
class ConfigController extends Controller {
    constructor(config, { versionService }) {
        super(config);
        this.versionService = versionService;
        const authenticationType = config.authentication && config.authentication.type;
        this.uiConfig = {
            ...config.ui,
            authenticationType,
            unleashUrl: config.server.unleashUrl,
            baseUriPath: config.server.baseUriPath,
            version: version_1.default,
        };
        this.get('/', this.getUIConfig);
    }
    async getUIConfig(req, res) {
        const config = this.uiConfig;
        if (this.versionService) {
            const versionInfo = this.versionService.getVersionInfo();
            res.json({ ...config, versionInfo });
        }
        else {
            res.json(config);
        }
    }
}
exports.default = ConfigController;
module.exports = ConfigController;
