// unleash-server/dist/test/config/test-config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTestConfig = void 0;
const deepmerge_1 = __importDefault(require("deepmerge"));
const option_1 = require("../../lib/types/option");
const no_logger_1 = __importDefault(require("../fixtures/no-logger"));
const create_config_1 = require("../../lib/create-config");
function mergeAll(objects) {
    return deepmerge_1.default.all(objects.filter(i => i));
}
function createTestConfig(config) {
    const testConfig = {
        getLogger: no_logger_1.default,
        authentication: { type: option_1.IAuthType.NONE, createAdminUser: false },
        server: { secret: 'really-secret' },
        session: {
            db: false,
        },
        versionCheck: { enable: false },
    };
    const options = mergeAll([testConfig, config]);
    return create_config_1.createConfig(options);
}
exports.createTestConfig = createTestConfig;
module.exports = {
    createTestConfig,
};
