// unleash-server/dist/lib/types/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.USER_DELETED = exports.USER_UPDATED = exports.USER_CREATED = exports.DB_POOL_UPDATE = exports.ADDON_CONFIG_DELETED = exports.ADDON_CONFIG_UPDATED = exports.ADDON_CONFIG_CREATED = exports.DROP_TAG_TYPES = exports.TAG_TYPE_IMPORT = exports.TAG_TYPE_UPDATED = exports.TAG_TYPE_DELETED = exports.TAG_TYPE_CREATED = exports.DROP_TAGS = exports.TAG_IMPORT = exports.TAG_DELETED = exports.TAG_CREATED = exports.DROP_PROJECTS = exports.PROJECT_IMPORT = exports.PROJECT_DELETED = exports.PROJECT_UPDATED = exports.PROJECT_CREATED = exports.CONTEXT_FIELD_DELETED = exports.CONTEXT_FIELD_UPDATED = exports.CONTEXT_FIELD_CREATED = exports.DROP_STRATEGIES = exports.STRATEGY_IMPORT = exports.STRATEGY_UPDATED = exports.STRATEGY_REACTIVATED = exports.STRATEGY_DEPRECATED = exports.STRATEGY_DELETED = exports.STRATEGY_CREATED = exports.DROP_FEATURES = exports.FEATURE_STALE_OFF = exports.FEATURE_STALE_ON = exports.FEATURE_UNTAGGED = exports.DROP_FEATURE_TAGS = exports.FEATURE_TAG_IMPORT = exports.FEATURE_TAGGED = exports.FEATURE_IMPORT = exports.FEATURE_REVIVED = exports.FEATURE_ARCHIVED = exports.FEATURE_UPDATED = exports.FEATURE_CREATED = exports.APPLICATION_CREATED = void 0;
exports.APPLICATION_CREATED = 'application-created';
exports.FEATURE_CREATED = 'feature-created';
exports.FEATURE_UPDATED = 'feature-updated';
exports.FEATURE_ARCHIVED = 'feature-archived';
exports.FEATURE_REVIVED = 'feature-revived';
exports.FEATURE_IMPORT = 'feature-import';
exports.FEATURE_TAGGED = 'feature-tagged';
exports.FEATURE_TAG_IMPORT = 'feature-tag-import';
exports.DROP_FEATURE_TAGS = 'drop-feature-tags';
exports.FEATURE_UNTAGGED = 'feature-untagged';
exports.FEATURE_STALE_ON = 'feature-stale-on';
exports.FEATURE_STALE_OFF = 'feature-stale-off';
exports.DROP_FEATURES = 'drop-features';
exports.STRATEGY_CREATED = 'strategy-created';
exports.STRATEGY_DELETED = 'strategy-deleted';
exports.STRATEGY_DEPRECATED = 'strategy-deprecated';
exports.STRATEGY_REACTIVATED = 'strategy-reactivated';
exports.STRATEGY_UPDATED = 'strategy-updated';
exports.STRATEGY_IMPORT = 'strategy-import';
exports.DROP_STRATEGIES = 'drop-strategies';
exports.CONTEXT_FIELD_CREATED = 'context-field-created';
exports.CONTEXT_FIELD_UPDATED = 'context-field-updated';
exports.CONTEXT_FIELD_DELETED = 'context-field-deleted';
exports.PROJECT_CREATED = 'project-created';
exports.PROJECT_UPDATED = 'project-updated';
exports.PROJECT_DELETED = 'project-deleted';
exports.PROJECT_IMPORT = 'project-import';
exports.DROP_PROJECTS = 'drop-projects';
exports.TAG_CREATED = 'tag-created';
exports.TAG_DELETED = 'tag-deleted';
exports.TAG_IMPORT = 'tag-import';
exports.DROP_TAGS = 'drop-tags';
exports.TAG_TYPE_CREATED = 'tag-type-created';
exports.TAG_TYPE_DELETED = 'tag-type-deleted';
exports.TAG_TYPE_UPDATED = 'tag-type-updated';
exports.TAG_TYPE_IMPORT = 'tag-type-import';
exports.DROP_TAG_TYPES = 'drop-tag-types';
exports.ADDON_CONFIG_CREATED = 'addon-config-created';
exports.ADDON_CONFIG_UPDATED = 'addon-config-updated';
exports.ADDON_CONFIG_DELETED = 'addon-config-deleted';
exports.DB_POOL_UPDATE = 'db-pool-update';
exports.USER_CREATED = 'user-created';
exports.USER_UPDATED = 'user-updated';
exports.USER_DELETED = 'user-deleted';
