// unleash-server/dist/lib/types/permissions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DELETE_API_TOKEN = exports.CREATE_API_TOKEN = exports.UPDATE_API_TOKEN = exports.UPDATE_ROLE = exports.READ_ROLE = exports.DELETE_ADDON = exports.UPDATE_ADDON = exports.CREATE_ADDON = exports.DELETE_PROJECT = exports.UPDATE_PROJECT = exports.CREATE_PROJECT = exports.DELETE_CONTEXT_FIELD = exports.UPDATE_CONTEXT_FIELD = exports.CREATE_CONTEXT_FIELD = exports.UPDATE_APPLICATION = exports.DELETE_STRATEGY = exports.UPDATE_STRATEGY = exports.CREATE_STRATEGY = exports.DELETE_FEATURE = exports.UPDATE_FEATURE = exports.CREATE_FEATURE = exports.CLIENT = exports.ADMIN = void 0;
exports.ADMIN = 'ADMIN';
exports.CLIENT = 'CLIENT';
exports.CREATE_FEATURE = 'CREATE_FEATURE';
exports.UPDATE_FEATURE = 'UPDATE_FEATURE';
exports.DELETE_FEATURE = 'DELETE_FEATURE';
exports.CREATE_STRATEGY = 'CREATE_STRATEGY';
exports.UPDATE_STRATEGY = 'UPDATE_STRATEGY';
exports.DELETE_STRATEGY = 'DELETE_STRATEGY';
exports.UPDATE_APPLICATION = 'UPDATE_APPLICATION';
exports.CREATE_CONTEXT_FIELD = 'CREATE_CONTEXT_FIELD';
exports.UPDATE_CONTEXT_FIELD = 'UPDATE_CONTEXT_FIELD';
exports.DELETE_CONTEXT_FIELD = 'DELETE_CONTEXT_FIELD';
exports.CREATE_PROJECT = 'CREATE_PROJECT';
exports.UPDATE_PROJECT = 'UPDATE_PROJECT';
exports.DELETE_PROJECT = 'DELETE_PROJECT';
exports.CREATE_ADDON = 'CREATE_ADDON';
exports.UPDATE_ADDON = 'UPDATE_ADDON';
exports.DELETE_ADDON = 'DELETE_ADDON';
exports.READ_ROLE = 'READ_ROLE';
exports.UPDATE_ROLE = 'UPDATE_ROLE';
exports.UPDATE_API_TOKEN = 'UPDATE_API_TOKEN';
exports.CREATE_API_TOKEN = 'CREATE_API_TOKEN';
exports.DELETE_API_TOKEN = 'DELETE_API_TOKEN';
