// unleash-server/dist/lib/types/user.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const gravatar_url_1 = __importDefault(require("gravatar-url"));
const joi_1 = __importDefault(require("joi"));
class User {
    constructor({ id, name, email, username, imageUrl, seenAt, loginAttempts, createdAt, }) {
        this.isAPI = false;
        if (!id) {
            throw new TypeError('Id is required');
        }
        if (!username && !email) {
            throw new TypeError('Username or Email is required');
        }
        joi_1.default.assert(email, joi_1.default.string().email(), 'Email');
        joi_1.default.assert(username, joi_1.default.string(), 'Username');
        joi_1.default.assert(name, joi_1.default.string(), 'Name');
        this.id = id;
        this.name = name;
        this.username = username;
        this.email = email;
        this.imageUrl = imageUrl || this.generateImageUrl();
        this.seenAt = seenAt;
        this.loginAttempts = loginAttempts;
        this.createdAt = createdAt;
    }
    generateImageUrl() {
        return gravatar_url_1.default(this.email || this.username, {
            size: 42,
            default: 'retro',
        });
    }
}
exports.default = User;
module.exports = User;
